ttnn/cpp/ttnn/operations/moreh/moreh_matmul/moreh_matmul.cpp

```diff
@@ -0,0 +1,42 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "moreh_matmul.hpp"
+
+#include "tt_dnn/op_library/moreh_helper_functions.hpp"
+#include "ttnn/operations/moreh/moreh_dot_op/moreh_dot.hpp"
+#include "ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_device_operation.hpp"
+
+namespace ttnn::operations::moreh::moreh_matmul {
+
+inline bool is_dot_forward(const Tensor& input, const Tensor& other, bool transpose_input, bool transpose_other) {
+    // TODO: non-4d support for dot.
+    if (input.get_legacy_shape().rank() != 4 || other.get_legacy_shape().rank() != 4) {
+        return false;
+    }
+
+    if (transpose_input || transpose_other) {
+        return false;
+    }
+
+    return tt::operations::primary::is_1d_tensor(input) && tt::operations::primary::is_1d_tensor(other) &&
+           tt::operations::primary::is_same_shape(input, other);
+}
+
+Tensor MorehMatmul::invoke(
+    const Tensor& input,
+    const Tensor& other,
+    bool transpose_input,
+    bool transpose_other,
+    const std::optional<Tensor>& output,
+    const std::optional<const Tensor> bias,
+    const std::optional<MemoryConfig>& output_mem_config,
```

Comment:
```suggestion
    const std::optional<MemoryConfig>& memory_config,
```

Reply:
This was also done in earlier merged ops, so I think it is fine. Only thing that absolutely needed to be `memory_config` is the Python interface